File: day8/day8.py
#!/usr/bin/env python3

import sys

def dec_diff(line):
    real = 0
    state = 0
    hexlen = 0
    for x in line:
        if state == 0:
            if x == "\\":
                state = 1
            elif x != "\"":
                real += 1
        elif state == 1:
            if x == "x":
                state = 2
                hexlen = 2
            elif x in "\\\"":
                real += 1
                state = 0
            else:
                state = 0
        elif state == 2:
            if x in "0123456789abcdef" and hexlen != 1:
                hexlen -= 1
            elif hexlen == 1:
                real += 1
                state = 0
            else:
                state = 0
        else:
            assert 0, "Unknown state {}".format(state)

    return len(line) - real

def enc_diff(line):
    reprlen = 0
    for x in line:
        if x in "\\\"":
            reprlen += 1
        reprlen += 1

    return reprlen + 2 - len(line)

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read()
    except OSError as e:
        print(e, file=sys.stderr)
        sys.exit(1)

    lines = txt.splitlines()
    print("Part1:", sum(map(dec_diff, lines)))
    print("Part2:", sum(map(enc_diff, lines)))

File: day8/day8.c
#include <assert.h>
#include <ctype.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

static size_t declength(const char *str)
{
	size_t real = 0, repr = 0;
	enum { NONE, ESCAPED, HEX1, HEX2 } state = NONE;
	for (; str[0] != 0 && str[0] != '\n'; str++)
	{
		repr++;
		if (state == NONE)
		{
			if (str[0] == '\\')
			{
				state = ESCAPED;
			}
			else if (str[0] != '"')
			{
				real++;
			}
		}
		else if (state == ESCAPED)
		{
			if (str[0] == 'x')
			{
				state = HEX1;
			}
			else if (str[0] == '\\' || str[0] == '"')
			{
				real++;
				state = NONE;
			}
			else
			{
				state = NONE;
			}
		}
		else if (state == HEX1)
		{
			if (isxdigit(str[0]))
			{
				state = HEX2;
			}
			else
			{
				real++;
				state = NONE;
			}
		}
		else if (state == HEX2)
		{
			if (isxdigit(str[0]))
			{
				real++;
			}
			state = NONE;
		}
		else
		{
			abort();
		}
	}
	return repr - real;
}

static size_t enclength(const char *s)
{
	size_t real = 0, repr = 0;
	for (; s[0] != 0 && s[0] != '\n'; s++)
	{
		real++;
		if (s[0] == '"' || s[0] == '\\')
		{
			repr += 2;
		}
		else
		{
			repr++;
		}
	}
	return repr + 2 - real;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fprintf(stderr, "Usage: %s <filename>\n", argv[0]);
		return 1;
	}

	FILE *input = fopen(argv[1], "rb");
	if (!input)
	{
		fprintf(stderr, "Cannot open %s\n", argv[1]);
		return 1;
	}

	assert(declength("\"\"") == 2-0);
	assert(declength("\"abc\"") == 5-3);
	assert(declength("\"aaa\\\"aaa\"") == 10-7);
	assert(declength("\"\\x27\"") == 6-1);

	assert(enclength("\"\"") == 6-2);
	assert(enclength("\"abc\"") == 9-5);
	assert(enclength("\"aaa\\\"aaa\"") == 16-10);
	assert(enclength("\"\\x27\"") == 11-6);

	size_t decdif = 0, encdif = 0;
	char *line;
	size_t linesize = 0;
	while (getline(&line, &linesize, input) != -1)
	{
		decdif += declength(line);
		encdif += enclength(line);
	}
	free(line);
	fclose(input);

	printf("Part1: %zu\n", decdif);
	printf("Part2: %zu\n", encdif);
	return 0;
}
